mathics/builtin/scoping.py

```diff
@@ -206,6 +206,142 @@ def apply(self, vars, expr, evaluation):
         result = new_expr.evaluate(evaluation)
         return result
 
+class Unique(Predefined):
+    """
+    <dl>
+    <dt>'Unique[]'
+        <dd>generates a new symbol and gives a name of the form '$number'.
+    <dt>'Unique[x]'
+        <dd>generates a new symbol and gives a name of the form 'x$number'.
+    <dt>'Unique[{x, y, ...}]'
+        <dd>generates a list of new symbols.
+    <dt>'Unique["xxx"]'
+        <dd>generates a new symbol and gives a name of the form 'xxxnumber'.
+    </dl>
+
+    Create a unique symbol with no particular name:
+    >> Unique[]
+     = $1
+
+    >> Unique[sym]
+     = sym$1
+
+    Create a unique symbol whose name begins with x:
+    >> Unique["x"]
+     = x2
+
+    #> Unique[]
+     = $3
+
+    #> Unique[{}]
+     = {}
+
+    #> Unique[{x, x}]
+     = {x$2, x$3}
+
+    Each use of Unique[symbol] increments $ModuleNumber:
+    >> {$ModuleNumber, Unique[x], $ModuleNumber}
+     = {4, x$4, 5}
+
+    Unique[symbol] creates symbols in the same way Module does:
+    >> {Module[{x}, x], Unique[x]}
+     = {x$5, x$6}
+
+    Unique with more arguments
+    >> Unique[{x, "s"}, Flat ^ Listable ^ Orderless]
+     : Flat ^ Listable ^ Orderless is not a known attribute.
+     = Unique[{x, s}, Flat ^ Listable ^ Orderless]
+
+    Unique call without symbol argument
+    >> Unique[x + y]
+     : x + y is not a symbol or a valid symbol name.
+     = Unique[x + y]
+
+    #> Unique[1]
+     : 1 is not a symbol or a valid symbol name.
+     = Unique[1]
+
+    #> Unique[{m, "s", n}, {Flat, Listable, Orderless}]
+     = {m$7, s4, n$8}
+     
+    #> Attributes[{m$7, s4, n$8}]
+     = {{Flat, Listable, Orderless}, {Flat, Listable, Orderless}, {Flat, Listable, Orderless}}
+
+    #> Unique[{x, "s", 1}, {Flat ^ Listable ^ Orderless}]
+     : 1 is not a symbol or a valid symbol name.
+     = Unique[{x, s, 1}, {Flat ^ Listable ^ Orderless}]
+
+    #> Unique[{"s"}, Flat]
+     = {s5}
+     
+    #> Attributes[s5]
+     = {Flat}
+    """
+
+    seq_number = 1
+
+    messages = {
+        'usym': '`1` is not a symbol or a valid symbol name.',
+        'argrx': 'Unique called with `1` arguments; 0 or 1 argument are expected.',
+        'attnf': '`1` is not a known attribute.',
+    }
+
+    attributes = ('Protected',)
+
+    rules = {
+        'Unique[x_Symbol]': 'Module[{x}, x]',
+    }
+
+    def apply(self, evaluation):
+        'Unique[]'
+
+        new_name = '$%d' % (self.seq_number)
+        self.seq_number += 1
+        return Symbol(new_name)
```

Comment:
We should check that the new symbol has not been defined yet. It's not unique otherwise ;)

```
In[1]:= Unique[]

Out[1]= $3

In[2]:= Unique[]

Out[2]= $4

In[3]:= $5 = 1

Out[3]= 1

In[4]:= Unique[]  (* $5 is skipped now *)

Out[4]= $6
```

Reply:
Fixed on https://github.com/mathics/Mathics/pull/696/commits/745a815bfb09f381f0ce36cb2ee6ea8a99d18d00.